File: src/main/java/link/pdfparser/PdfParser.java

/*
    This file is part of the iText (R) project.
    Copyright (c) 1998-2020 iText Group NV
    Authors: iText Software.

    For more information, please contact iText Software at this address:
    sales@itextpdf.com
 */
/**
 * Example written by Bruno Lowagie in answer to:
 * http://stackoverflow.com/questions/24506830/can-we-use-text-extraction-strategy-after-applying-location-extraction-strategy
 */
package link.pdfparser;

import com.itextpdf.kernel.font.PdfFont;
import com.itextpdf.kernel.geom.Rectangle;
import com.itextpdf.kernel.pdf.PdfDocument;
import com.itextpdf.kernel.pdf.PdfReader;
import com.itextpdf.kernel.pdf.canvas.parser.EventType;
import com.itextpdf.kernel.pdf.canvas.parser.PdfCanvasProcessor;
import com.itextpdf.kernel.pdf.canvas.parser.data.IEventData;
import com.itextpdf.kernel.pdf.canvas.parser.data.TextRenderInfo;
import com.itextpdf.kernel.pdf.canvas.parser.filter.IEventFilter;
import com.itextpdf.kernel.pdf.canvas.parser.filter.TextRegionEventFilter;
import com.itextpdf.kernel.pdf.canvas.parser.listener.FilteredEventListener;
import com.itextpdf.kernel.pdf.canvas.parser.listener.IEventListener;
import com.itextpdf.kernel.pdf.canvas.parser.listener.LocationTextExtractionStrategy;
import com.itextpdf.kernel.pdf.canvas.parser.listener.SimpleTextExtractionStrategy;

import java.io.File;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.FileOutputStream;
import java.io.Writer;
import java.io.BufferedWriter;
import java.util.Set;

import static com.itextpdf.kernel.pdf.canvas.parser.EventType.*;

public class PdfParser {
    public static final String DEST = "./Gnetzwerk_a1-b1_unregelmaessigeverben.txt";

    public static final String SRC = "/Users/oleksandr.linkevych/Downloads/netzwerk_a1-b1_unregelmaessigeverben (1).pdf";

    public static void main(String[] args) throws IOException {
        new PdfParser().manipulatePdf(DEST);
    }

    protected void manipulatePdf(String dest) throws IOException {
        PdfDocument pdfDoc = new PdfDocument(new PdfReader(SRC));

//        pdfDoc.getPage(10)

//        CustomAlignFilter fontFilter = new CustomAlignFilter();
        FilteredEventListener listener = new FilteredEventListener();

        // Create a text extraction renderer
        PdfEventListener extractionStrategy = listener
                .attachEventListener(new PdfEventListener()/*, fontFilter*/);

        // Note: If you want to re-use the PdfCanvasProcessor, you must call PdfCanvasProcessor.reset()
        PdfCanvasProcessor parser = new PdfCanvasProcessor(listener);
        for (int i = 1; i <= pdfDoc.getNumberOfPages(); i++) {
            parser.processPageContent(pdfDoc.getPage(i));
        }


        // Get the resultant text after applying the custom filter
//        String actualText = extractionStrategy.getResultantText();

//        System.out.println(actualText);

        pdfDoc.close();

        // See the resultant text in the console
//        System.out.println(actualText);

//        try (Writer writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(dest)))) {
//            writer.write(actualText);
//        }
    }

    class PdfEventListener implements IEventListener {

        @Override
        public void eventOccurred(IEventData data, EventType type) {
            TextRenderInfo textRenderInfo = (TextRenderInfo) data;
            System.out.println(textRenderInfo.getText() + " - " + textRenderInfo.getBaseline().getBoundingRectangle().getY());
        }

        @Override
        public Set<EventType> getSupportedEvents() {
            return Set.of(RENDER_TEXT);
        }
    }


}

File: src/main/scala/link/german/gender/PdfParser.scala
package link.german.gender

import java.util

import com.itextpdf.kernel.pdf.canvas.parser.EventType.RENDER_TEXT
import com.itextpdf.kernel.pdf.{PdfDocument, PdfReader}
import com.itextpdf.kernel.pdf.canvas.parser.{EventType, PdfCanvasProcessor}
import com.itextpdf.kernel.pdf.canvas.parser.data.{IEventData, TextRenderInfo}
import com.itextpdf.kernel.pdf.canvas.parser.listener.{FilteredEventListener, IEventListener}

import scala.collection.JavaConverters._
import scala.collection.mutable

object PdfParser extends App {

  val SRC = "/Users/oleksandr.linkevych/Downloads/netzwerk_a1-b1_unregelmaessigeverben (1).pdf"

  val pdfDoc = new PdfDocument(new PdfReader(SRC))

  //        pdfDoc.getPage(10)

  //        CustomAlignFilter fontFilter = new CustomAlignFilter();
  val listener = new FilteredEventListener

  // Create a text extraction renderer
  val extractionStrategy = listener.attachEventListener(new PdfEventListener)

  // Note: If you want to re-use the PdfCanvasProcessor, you must call PdfCanvasProcessor.reset()
  val parser = new PdfCanvasProcessor(listener)
  for (i <- 1 to pdfDoc.getNumberOfPages) {
    parser.processPageContent(pdfDoc.getPage(i))
  }

  println(extractionStrategy.getText)

  class PdfEventListener extends IEventListener {

    case class TextBlock(text: String, left: Float, top: Float, right: Float, bottom: Float)

    val dataSeq: mutable.Buffer[TextBlock] = mutable.Buffer[TextBlock]()

    def cluster[T](seq: Seq[T], join: PartialFunction[(T, T), T]): Seq[T] = {
      seq.foldLeft(Seq[T]()){
        case (Seq(), t) => Seq(t)
        case (buf :+ last, t) => buf ++ join.andThen(Seq(_)).applyOrElse(last -> t, (_: (T, T)) => Seq(last, t))
      }
    }

    def getText: String = dataSeq
      .groupBy(_.top)
      .mapValues(v => cluster[TextBlock](v.sortBy(_.left), {
        case (v1, v2) if Math.abs(v2.left - v1.right) < 1 =>
          TextBlock(v1.text + v2.text, v1.left, v1.top, v2.right, Math.max(v1.bottom, v2.bottom))
      })).toSeq.sortBy(_._1).map(_._2).map(_.map(_.text.trim).mkString("|")).mkString("\n")

    override def eventOccurred(data: IEventData, `type`: EventType): Unit = {
      val info = data.asInstanceOf[TextRenderInfo]
      dataSeq += TextBlock(info.getText,
        info.getBaseline.getBoundingRectangle.getLeft,
        info.getBaseline.getBoundingRectangle.getTop,
        info.getBaseline.getBoundingRectangle.getRight,
        info.getBaseline.getBoundingRectangle.getBottom
      )
    }

    override def getSupportedEvents: util.Set[EventType] = Set(RENDER_TEXT).asJava
  }


}
